kernel/src/process.rs

```diff
@@ -585,19 +635,23 @@ impl<C: Chip> ProcessType for Process<'a, C> {
                 // process faulted. Panic and print status
                 panic!("Process {} had a fault", self.process_name);
             }
-            FaultResponse::Restart => {
-                // Remove the tasks that were scheduled for the app from the
-                // amount of work queue.
-                let tasks_len = self.tasks.map_or(0, |tasks| tasks.len());
-                for _ in 0..tasks_len {
-                    self.kernel.decrement_work();
+            FaultResponse::Restart(restart_policy) => {
+                // Start with the generic terminate operations. This frees state
+                // for this process and removes any pending tasks from the
+                // scheduler's queue.
+                self.terminate();
+
+                // Decide what to do with this process. Should it be restarted?
+                // Or should we leave it in a stopped & faulted state? If the
+                // process is faulting too often we might not want to restart.
+                let should_restart = restart_policy.evaluate(self);
+
+                // If we are not going to restart the process then we can just
+                // leave it in the stopped faulted state.
+                if !should_restart {
+                    return;
```

Comment:
So if we change the name of the trait method, this would be more readable as either:

```rust
if restart_policy.should_restart(self) {
    return;
}
```

or:

```rust
if let FaultBehavior::Restart = restart_policy.fault_behavior(self) {
    return;
}
```